### DISCUSSION TITLE
Simple getting-started questions

### DISCUSSION THREAD

**Discussion by blakemcbride** [antlr/antlr4#4712] [General]
Greetings,

Although I've known of ANTLR since the C days, I've never used it before.  I've built lots of parsers by hand in the interim.  This time, I'm hoping not to work so hard.

I wrote a grammar that seems to work.  I am running it as follows:
```
    private static void test2() {
        String pgm = "if SQ-44 > 77 then display \"Some sort of error\"";
        CharStream input = CharStreams.fromString(pgm);
        SurveyLogicLexer lexer = new SurveyLogicLexer(input);
        CommonTokenStream tokens = new CommonTokenStream(lexer);
        SurveyLogicParser parser = new SurveyLogicParser(tokens);

        ParseTree tree = parser.init();
        String x = tree.toStringTree(parser);
        System.out.println(tree.toStringTree(parser));
        //ParseTreeWalker walker = new ParseTreeWalker();
        //walker.walk(new SurveyLogicBaseListener(), tree);
    }
```
Here is what I need:

1. The above code appears to print to the console syntax errors (before the println).  I'd rather catch an exception or check a return (along with an error string) to handle it myself.  How can I do that?

2. I think I'd rather traverse the tree rather than using callbacks.  Is that a bad idea?  How can I do that?

Thanks!

Blake McBride

**Comment by kaby76** [comment]
To change the output on error, [remove all listeners](https://github.com/antlr/grammars-v4/blob/6d13b1068d0fc9eba30a3f291fe62026fbc71c2f/_scripts/templates/Java/st.Test.java#L175-L176) and [add your listener](https://github.com/antlr/grammars-v4/blob/6d13b1068d0fc9eba30a3f291fe62026fbc71c2f/_scripts/templates/Java/st.Test.java#L177-L178). In your listener code, do what you need to do to [change how it looks or where to output](https://github.com/antlr/grammars-v4/blob/6d13b1068d0fc9eba30a3f291fe62026fbc71c2f/_scripts/templates/Java/st.ErrorListener.java#L32-L40).

For your tree walker, you don't need to use any of the generated listener or visitor classes that are generated, or the [tree walker](https://github.com/antlr/antlr4/blob/2a7904a595479954ccfb1c78124fc3d7f5bebcb1/runtime/Java/src/org/antlr/v4/runtime/tree/IterativeParseTreeWalker.java#L22-L80) or [visitor code](https://github.com/antlr/antlr4/blob/2a7904a595479954ccfb1c78124fc3d7f5bebcb1/runtime/Java/src/org/antlr/v4/runtime/tree/AbstractParseTreeVisitor.java#L9-L141). You can write your own tree walker code. Or, you can use a higher-level [tree pattern matcher using XPath in Antlr](https://github.com/antlr/antlr4/tree/2a7904a595479954ccfb1c78124fc3d7f5bebcb1/runtime/Java/src/org/antlr/v4/runtime/tree/xpath), or [external](https://github.com/kaby76/Trash/tree/main/src/trxgrep#readme).

**Comment by blakemcbride** [comment]
Thanks for the help!  Just FYI, chatGPT gives really good examples!

**Reply by kaby76** [reply]
ChatGPT will give examples, yes. For simple problems, it sometimes gives correct answers. But, for more complicated problems, Chat is more often wrong than right because the result is not actually tested. https://github.com/antlr/antlr4/discussions/4127. It can help for simple "one shot" problems, which you can then follow up on by testing yourself. However, problem descriptions are not formal. So, the problem statement cannot be reliably reworked with different initial conditions, or the problem statement ported to an entirely different LLM. It's the same reason why math problems can't be solved via Chat. https://techcrunch.com/2024/10/02/why-is-chatgpt-so-bad-at-math/. Memorizing a ton of grammars does not give it the reasoning power to understand why a grammar isn't working or why it does not perform well.

**Reply by blakemcbride** [reply]
I've found the new 01 model incredible and quite a bit smarter than the old 4o model.

**Reply by kaby76** [reply]
Here is an update to the question "Please scrape the EBNF at https://flang.llvm.org/docs/f2018-grammar.html. Please give the complete grammar." with the new LLM. https://chatgpt.com/share/67001b53-6450-8007-9516-b1655b9d7c7f. It still does not give the complete answer.

When I ask to give the definition for "rename-list", it gives that correctly (`rename-list    ::= rename { , rename }`), but defines "rename" incorrectly, when it could have just left it there, or given the EBNF of the rule as something like `rename ::= local-name => use-name | OPERATOR ( local-defined-operator ) => OPERATOR ( use-defined-operator )`. Imstead. ot follows up with all sorts of BS with regards to "use-name", completely irrelevant.

**Comment by jimidle** [comment]
I was just going to say the same. Much better results. The new canvas stuff may do even better but I have tested itOn Oct 4, 2024, at 11:08, Blake McBride ***@***.***> wrote:﻿
I've found the new 01 model incredible and quite a bit smarter than the old 4o model.

**Comment by jimidle** [comment]
I think it still does not  scrape URls Ken. But the canvas stuff is supposed to do that I think. On Oct 4, 2024, at 11:57, Ken Domino ***@***.***> wrote:﻿
Here is an update to the question "Please scrape the EBNF at https://flang.llvm.org/docs/f2018-grammar.html. Please give the complete grammar." with the new LLM. https://chatgpt.com/share/67001b53-6450-8007-9516-b1655b9d7c7f. It still does not give the complete answer.
When I ask to give the definition for "rename-list", it gives that correctly (rename-list    ::= rename { , rename }), but defines "rename" incorrectly, when it could have just left it there, or given the EBNF of the rule as something like rename ::= local-name => use-name | OPERATOR ( local-defined-operator ) => OPERATOR ( use-defined-operator ). Imstead. ot follows up with all sorts of BS with regards to "use-name", completely irrelevant.

**Reply by kaby76** [reply]
I'll gave that a try. Chatgpt 4o-preview seems to be able to follow links. I know it's something that they want to offer.

**Comment by jimidle** [comment]
I can’t really think of making it much higher level to be honest. There are options and so on at the lexer level and parser level. I think trying to treat the lever and parser as one will lead to more confusion about the order of operation than we already get. Parsing is fundamentally not the  trivial problem it seems to be when you want can efficient parser. We would do better to spend time on better documents about performance analysis and ambiguities. Th ease questions come up very often and by that time people have assembled a bunch of random rules without understanding such things. I’d even go so far as to say that the default options and reporting should print all tracing etc and you get to turn them off rather than on. Upgrading would require a change or we could maybe deprecate existing methods and document new ones for new people. Just done random thoughtsJimOn Oct 4, 2024, at 11:18, Blake McBride ***@***.***> wrote:﻿
Incidentally, while working with the latest ANTLR, I've come to the following opinion.  ANTLR provides a very powerful and functional toolkit.  However, a higher-level API is sorely needed (without getting rid of the toolkit API).  A great deal of, effectively, boiler-plate code could be combined to make this higher-level API.  It would make ANTLR far, far easier to use.
Some examples:
Why treat the lexer and parser as two different entities throughout when they're interrelated.  Tread them as a unit.
Except for testing, why would anyone want errors printed to the console?  Put them in a linked list and then provide a simple utility to dump them to the console when testing rather than the other way around.
Many, many more instances of stuff like this.  You could easily create a high-level API that would make using ANTLR trivial rather than a pretty complex project.
